cmd/evm/staterunner.go

```diff
@@ -35,157 +35,123 @@ import (
-	for key, test := range testsByName {
-		for _, st := range test.Subtests() {
-			if idx != -1 && st.Index != idx {
-				continue
-			}
-			if fork != "" && st.Fork != fork {
-				continue
-			}
-			res = append(res, stateTestCase{name: key, st: st, test: test})
-		}
-	}
-	return res
-}
-
 // runStateTest loads the state-test given by fname, and executes the test.
-func runStateTest(ctx *cli.Context, fname string, cfg vm.Config, dump bool, bench bool) error {
+func runStateTest(ctx *cli.Context, fname string) ([]testResult, error) {
 	src, err := os.ReadFile(fname)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	var testsByName map[string]tests.StateTest
 	if err := json.Unmarshal(src, &testsByName); err != nil {
-		return err
+		return nil, fmt.Errorf("unable to read test file %s: %w", fname, err)
 	}
 
-	matchingTests := collectMatchedSubtests(ctx, testsByName)
+	cfg := vm.Config{Tracer: tracerFromFlags(ctx)}
+	re, err := regexp.Compile(ctx.String(RunFlag.Name))
+	if err != nil {
+		return nil, fmt.Errorf("invalid regex -%s: %v", RunFlag.Name, err)
+	}
 
 	// Iterate over all the tests, run them and aggregate the results
-	var results []StatetestResult
-	for _, test := range matchingTests {
-		// Run the test and aggregate the result
-		result := &StatetestResult{Name: test.name, Fork: test.st.Fork, Pass: true}
-		test.test.Run(test.st, cfg, false, rawdb.HashScheme, func(err error, tstate *tests.StateTestState) {
-			var root common.Hash
-			if tstate.StateDB != nil {
-				root = tstate.StateDB.IntermediateRoot(false)
-				result.Root = &root
-				fmt.Fprintf(os.Stderr, "{\"stateRoot\": \"%#x\"}\n", root)
-				if dump { // Dump any state to aid debugging
-					cpy, _ := state.New(root, tstate.StateDB.Database())
-					dump := cpy.RawDump(nil)
-					result.State = &dump
-				}
+	results := make([]testResult, 0, len(testsByName))
+	for key, test := range testsByName {
+		if !re.MatchString(key) {
+			continue
+		}
+		for i, st := range test.Subtests() {
+			if idx := ctx.Int(idxFlag.Name); idx != -1 && idx != i {
+				// If specific index requested, skip all tests that do not match.
+				continue
 			}
-			if err != nil {
-				// Test failed, mark as so
-				result.Pass, result.Error = false, err.Error()
+			if fork := ctx.String(forkFlag.Name); fork != "" && st.Fork != fork {
+				// If specific fork requested, skip all tests that do not match.
+				continue
 			}
-		})
-		if bench {
-			_, stats, _ := timedExec(true, func() ([]byte, uint64, error) {
-				_, _, gasUsed, _ := test.test.RunNoVerify(test.st, cfg, false, rawdb.HashScheme)
-				return nil, gasUsed, nil
+			// Run the test and aggregate the result
+			result := &testResult{Name: key, Fork: st.Fork, Pass: true}
+			test.Run(st, cfg, false, rawdb.HashScheme, func(err error, state *tests.StateTestState) {
+				var root common.Hash
+				if state.StateDB != nil {
+					root = state.StateDB.IntermediateRoot(false)
+					result.Root = &root
+					// Dump any state to aid debugging.
```

Comment:
Did you just drop the `fmt.Fprintf(os.Stderr` - output of the stateRoot?

Reply:
The state root is recorded in the result and output via `report(..)`.